Cryptic error on Map/Set/WeakMap/WeakSet

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.4.0-dev.20190202

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** Map, WeakMap, Set

**Code**
```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
const instances = new WeakMap<HTMLElement, string>();
```
Feel free to replace `WeakMap` by `Map`, or `WeakSet`/`Set` (but with one type). The problem will persist.

**Expected behavior:**
No error. This code is perfectly fine. Without the types, it would be valid javascript.

**Actual behavior:**
```
test.ts:1:23 - error TS2693: 'WeakMap' only refers to a type, but is being used as a value here.

1 const instances = new WeakMap<HTMLElement, string>();
                        ~~~~~~~
```
The code does compile & function absolutely fine. As expected, because there's nothing wrong with it.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
[Playground doesn't show the error, but here it is anyway](https://www.typescriptlang.org/play/#src=const%20instances%20%3D%20new%20WeakMap%3CHTMLElement%2C%20string%3E()%3B)

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
None that I coud find.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -1716,12 +1716,25 @@ namespace ts {
                 }
                 const symbol = resolveSymbol(resolveName(errorLocation, name, SymbolFlags.Type & ~SymbolFlags.Value, /*nameNotFoundMessage*/undefined, /*nameArg*/ undefined, /*isUse*/ false));
                 if (symbol && !(symbol.flags & SymbolFlags.NamespaceModule)) {
-                    const message = (name === "Promise" || name === "Symbol")
+                    const message = isES2015OrLaterConstructorName(name)
                         ? Diagnostics._0_only_refers_to_a_type_but_is_being_used_as_a_value_here_Do_you_need_to_change_your_target_library_Try_changing_the_lib_compiler_option_to_es2015_or_later
                         : Diagnostics._0_only_refers_to_a_type_but_is_being_used_as_a_value_here;
                     error(errorLocation, message, unescapeLeadingUnderscores(name));
                     return true;
                 }
+            }
+            return false;
+        }
+
+        function isES2015OrLaterConstructorName(n: __String) {
+            switch (n) {
+                case "Promise":
+                case "Symbol":
+                case "Map":
+                case "WeakMap":
+                case "Set":
+                case "WeakSet":
+                    return true;
             }
             return false;
         }

--- a/tests/baselines/reference/forwardDeclaredCommonTypes01.errors.txt
+++ b/tests/baselines/reference/forwardDeclaredCommonTypes01.errors.txt
@@ -1,0 +1,40 @@
+tests/cases/compiler/forwardDeclaredCommonTypes01.ts(9,9): error TS2585: 'Promise' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+tests/cases/compiler/forwardDeclaredCommonTypes01.ts(10,9): error TS2585: 'Symbol' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+tests/cases/compiler/forwardDeclaredCommonTypes01.ts(10,17): error TS2585: 'Symbol' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+tests/cases/compiler/forwardDeclaredCommonTypes01.ts(11,9): error TS2585: 'Map' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+tests/cases/compiler/forwardDeclaredCommonTypes01.ts(12,9): error TS2585: 'WeakMap' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+tests/cases/compiler/forwardDeclaredCommonTypes01.ts(13,9): error TS2585: 'Set' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+tests/cases/compiler/forwardDeclaredCommonTypes01.ts(14,9): error TS2585: 'WeakSet' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+
+
+==== tests/cases/compiler/forwardDeclaredCommonTypes01.ts (7 errors) ====
+    interface Promise<T> {}
+    interface Symbol {}
+    interface Map<K, V> {}
+    interface WeakMap<K extends object, V> {}
+    interface Set<T> {}
+    interface WeakSet<T extends object> {}
+    
+    (function() {
+        new Promise;
+            ~~~~~~~
+!!! error TS2585: 'Promise' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+        new Symbol; Symbol();
+            ~~~~~~
+!!! error TS2585: 'Symbol' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+                    ~~~~~~
+!!! error TS2585: 'Symbol' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+        new Map;
+            ~~~
+!!! error TS2585: 'Map' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+        new WeakMap;
+            ~~~~~~~
+!!! error TS2585: 'WeakMap' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+        new Set;
+            ~~~
+!!! error TS2585: 'Set' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+        new WeakSet;
+            ~~~~~~~
+!!! error TS2585: 'WeakSet' only refers to a type, but is being used as a value here. Do you need to change your target library? Try changing the `lib` compiler option to es2015 or later.
+    });
+    

--- a/tests/baselines/reference/forwardDeclaredCommonTypes01.js
+++ b/tests/baselines/reference/forwardDeclaredCommonTypes01.js
@@ -1,0 +1,28 @@
+//// [forwardDeclaredCommonTypes01.ts]
+interface Promise<T> {}
+interface Symbol {}
+interface Map<K, V> {}
+interface WeakMap<K extends object, V> {}
+interface Set<T> {}
+interface WeakSet<T extends object> {}
+
+(function() {
+    new Promise;
+    new Symbol; Symbol();
+    new Map;
+    new WeakMap;
+    new Set;
+    new WeakSet;
+});
+
+
+//// [forwardDeclaredCommonTypes01.js]
+(function () {
+    new Promise;
+    new Symbol;
+    Symbol();
+    new Map;
+    new WeakMap;
+    new Set;
+    new WeakSet;
+});

--- a/tests/baselines/reference/forwardDeclaredCommonTypes01.symbols
+++ b/tests/baselines/reference/forwardDeclaredCommonTypes01.symbols
@@ -1,0 +1,35 @@
+=== tests/cases/compiler/forwardDeclaredCommonTypes01.ts ===
+interface Promise<T> {}
+>Promise : Symbol(Promise, Decl(lib.es5.d.ts, --, --), Decl(forwardDeclaredCommonTypes01.ts, 0, 0))
+>T : Symbol(T, Decl(lib.es5.d.ts, --, --), Decl(forwardDeclaredCommonTypes01.ts, 0, 18))
+
+interface Symbol {}
+>Symbol : Symbol(Symbol, Decl(lib.es5.d.ts, --, --), Decl(forwardDeclaredCommonTypes01.ts, 0, 23))
+
+interface Map<K, V> {}
+>Map : Symbol(Map, Decl(forwardDeclaredCommonTypes01.ts, 1, 19))
+>K : Symbol(K, Decl(forwardDeclaredCommonTypes01.ts, 2, 14))
+>V : Symbol(V, Decl(forwardDeclaredCommonTypes01.ts, 2, 16))
+
+interface WeakMap<K extends object, V> {}
+>WeakMap : Symbol(WeakMap, Decl(forwardDeclaredCommonTypes01.ts, 2, 22))
+>K : Symbol(K, Decl(forwardDeclaredCommonTypes01.ts, 3, 18))
+>V : Symbol(V, Decl(forwardDeclaredCommonTypes01.ts, 3, 35))
+
+interface Set<T> {}
+>Set : Symbol(Set, Decl(forwardDeclaredCommonTypes01.ts, 3, 41))
+>T : Symbol(T, Decl(forwardDeclaredCommonTypes01.ts, 4, 14))
+
+interface WeakSet<T extends object> {}
+>WeakSet : Symbol(WeakSet, Decl(forwardDeclaredCommonTypes01.ts, 4, 19))
+>T : Symbol(T, Decl(forwardDeclaredCommonTypes01.ts, 5, 18))
+
+(function() {
+    new Promise;
+    new Symbol; Symbol();
+    new Map;
+    new WeakMap;
+    new Set;
+    new WeakSet;
+});
+

--- a/tests/baselines/reference/forwardDeclaredCommonTypes01.types
+++ b/tests/baselines/reference/forwardDeclaredCommonTypes01.types
@@ -1,0 +1,40 @@
+=== tests/cases/compiler/forwardDeclaredCommonTypes01.ts ===
+interface Promise<T> {}
+interface Symbol {}
+interface Map<K, V> {}
+interface WeakMap<K extends object, V> {}
+interface Set<T> {}
+interface WeakSet<T extends object> {}
+
+(function() {
+>(function() {    new Promise;    new Symbol; Symbol();    new Map;    new WeakMap;    new Set;    new WeakSet;}) : () => void
+>function() {    new Promise;    new Symbol; Symbol();    new Map;    new WeakMap;    new Set;    new WeakSet;} : () => void
+
+    new Promise;
+>new Promise : any
+>Promise : any
+
+    new Symbol; Symbol();
+>new Symbol : any
+>Symbol : any
+>Symbol() : any
+>Symbol : any
+
+    new Map;
+>new Map : any
+>Map : any
+
+    new WeakMap;
+>new WeakMap : any
+>WeakMap : any
+
+    new Set;
+>new Set : any
+>Set : any
+
+    new WeakSet;
+>new WeakSet : any
+>WeakSet : any
+
+});
+

--- a/tests/cases/compiler/forwardDeclaredCommonTypes01.ts
+++ b/tests/cases/compiler/forwardDeclaredCommonTypes01.ts
@@ -1,0 +1,18 @@
+// @lib: es5
+// @target: es5
+
+interface Promise<T> {}
+interface Symbol {}
+interface Map<K, V> {}
+interface WeakMap<K extends object, V> {}
+interface Set<T> {}
+interface WeakSet<T extends object> {}
+
+(function() {
+    new Promise;
+    new Symbol; Symbol();
+    new Map;
+    new WeakMap;
+    new Set;
+    new WeakSet;
+});

